File: src/univ_prop/api/propagator.py
import numpy as np
from numpy.linalg import norm
from ..pyuniversal.kepler import solve_chi
from ..pyuniversal.coef import f, g, df, dg
from ..pyuniversal.stumpff import stumpff_C, stumpff_S
from typing import Tuple

class UniversalPropagator:
    def __init__(self, mu: float, name: str):
        self.name = name
        self.mu = mu
        self.tol = 1e-8
    
    def propagate(
        self,
        Position: list[float] | np.ndarray,
        Velocity: list[float] | np.ndarray,
        dt: float
    ) -> Tuple[np.ndarray, np.ndarray]:
        """
        Propagate the orbit forward by one time step using universal variable formulation.

        :param Position: Initial position vector [x, y, z] in kilometers.
        :type Position: list[float] or numpy.ndarray
        :param Velocity: Initial velocity vector [vx, vy, vz] in km/s.
        :type Velocity: list[float] or numpy.ndarray
        :param dt: Time step duration in seconds.
        :type dt: float

        :return: A tuple containing:
                 - r_vec (numpy.ndarray): Propagated position vector in km.
                 - v_vec (numpy.ndarray): Propagated velocity vector in km/s.
        :rtype: tuple[numpy.ndarray, numpy.ndarray]
        """

        mu = self.mu
        tol = self.tol

        r0_vec = np.array(Position)
        v0_vec = np.array(Velocity)
        r0 = norm(r0_vec)
        v0 = norm(v0_vec)
        vr = np.dot(r0_vec, v0_vec) / r0

        alpha = 2 / r0 - v0**2 / mu

        try:
            chi = solve_chi(mu, alpha, r0_vec, vr, dt, tol=tol, verbose=False)
        except RuntimeError as e:
            print("Propagation failed:", e)
            raise

        z = alpha * chi**2
        C = stumpff_C(z)
        S = stumpff_S(z)

        _f = f(chi,r0,C)
        _g = g(dt,chi,mu,S)

        r_vec = _f * r0_vec + _g * v0_vec
        r = norm(r_vec)

        fdot = df(chi,r0,r,S,z,mu)
        gdot = dg(chi,r,C)

        v_vec = fdot * r0_vec + gdot * v0_vec
        
        return r_vec, v_vec

File: src/univ_prop/src_/propagator.cpp
#include "kepler.h"
#include "stumpff.h"
#include "coef.h"
#include "propagator.h"

#include <iostream>
#include <vector>
#include <cmath>
#include <stdexcept>
#include <string>
#include <utility>

namespace univ_prop {
    class UniversalPropagator {
    public:
        std::string name;
        double mu;
        double tol;

        UniversalPropagator(double mu_, const std::string& name_)
            : name(name_), mu(mu_), tol(1e-8) {}

        std::pair<std::vector<double>, std::vector<double>> propagate(
            const std::vector<double>& Position, 
            const std::vector<double>& Velocity, 
            double time_step) 
        {
            const std::vector<double>& r0_vec = Position;
            const std::vector<double>& v0_vec = Velocity;

            double r0 = norm(Position);
            double v0 = norm(Velocity);
            double vr = (r0_vec[0] * r0_vec[0] + r0_vec[1] * v0_vec[1] + v0_vec[2] * v0_vec[2]) / r0;

            double alpha = 2.0 / (r0) - (v0 * v0) / mu;

            double chi;
            try {
                chi = solve_chi(mu, alpha, Position, vr, time_step, tol);
            } catch (const std::runtime_error& e){
                std::cerr << "Propagation failed: " << e.what() << std::endl;
                throw;
            }

            double z = alpha * chi * chi;
            double C = stumpff_C(z);
            double S = stumpff_S(z);

            double _f = f(chi, r0, C);
            double _g = g(chi, mu, S, time_step);

            std::vector<double> r_vec(3);
            for (int i =0; i < 3; ++i) {
                r_vec[i] = r0_vec[i] * _f + v0_vec[i] * _g;
            }
            double r = norm(r_vec);

            double _d_f = d_f(chi, r0, r, S, z, mu);
            double _d_g = d_g(chi, r, C);

            std::vector<double> v_vec(3);
            for (int i = 0; i < 3; ++i) {
                v_vec[i] = r0_vec[i] * _d_f + v0_vec[i] * _d_g;
            }

            return {r_vec, v_vec};
        }
    };
}